### ISSUE TITLE
Extract direction of effect from variant annotations

### ISSUE THREAD

**Issue by apriltuesday** [EBIvariation/opentargets-pharmgkb#20]
PGKB says they now provide the deconstructed sentences for the clinical annotations in their downloads, so it should be possible to pull out directionality related to the genotype-phenotype association.

**Comment by apriltuesday** [comment]
I have finally located this, it is in [variant annotations](https://api.pharmgkb.org/v1/download/file/data/variantAnnotations.zip) which looks to be very useful - one example with a few columns excluded:

| Variant/Haplotypes | Gene    | Drug(s)                 | Phenotype Category | Sentence                                          | Alleles | isPlural | Is/Is Not associated | Direction of effect | Functional terms | Gene/gene product | When treated with/exposed to/when assayed with | Multiple drugs And/or | Cell type      | Comparison Allele(s) or Genotype(s) |
| ------------------ | ------- | ----------------------- | ------------------ | ------------------------------------------------- | ------- | -------- | -------------------- | ------------------- | ---------------- | ----------------- | ---------------------------------------------- | --------------------- | -------------- | ----------------------------------- |
| rs1349931378       | CYP2C19 | mephenytoin, omeprazole | Metabolism/PK      | Allele A is associated with decreased activity... | A       | Is       | Associated with      | decreased           | activity of      | CYP2C19           | when assayed with                              | or                    | in 293FT cells | G                                   |

**Comment by apriltuesday** [comment]
The following is a summary of the notebook [here](https://nbviewer.org/github/apriltuesday/opentargets-pharmgkb/blob/doe/data-exploration/direction-of-effect.ipynb?flush_cache=true), looking into variant annotation tables generally and direction of effect more specifically.
@tcezard @dsuveges @ireneisdoomed @tskir Please have a look if you like and leave your comments/questions, otherwise we'll go through this notebook at a subsequent meeting.

#### Overview

PharmGKB provides 3 variant annotation tables, described in the readme as follows:
>* **var_pheno_ann.tsv**: Contains associations in which the variant affects a phenotype, with or without drug information. [13,517 rows in the 2024-05 data dump]
>* **var_drug_ann.tsv**: Contains associations in which the variant affects a drug dose, response, metabolism, etc. [11,901 rows]
>* **var_fa_ann.tsv**: Contains in vitro and functional analysis-type associations. [2,009 rows]

These variant annotations (plus drug labels/guidelines, not covered here) provide evidence for the clinical annotations, connected via an evidence ID.  The three types of variant annotations have different but overlapping columns; in general, each row describes an assertion made by a publication (PMID) about the effect of one or more allele/genotypes. See examples [here](https://nbviewer.org/github/apriltuesday/opentargets-pharmgkb/blob/doe/data-exploration/direction-of-effect.ipynb?flush_cache=true#Example-clinical-annotation) and some breakdown of how the information is provided [here](https://nbviewer.org/github/apriltuesday/opentargets-pharmgkb/blob/doe/data-exploration/direction-of-effect.ipynb?flush_cache=true#Sentence-breakdown).

Taken together, the variant annotations provide evidence for all clinical annotations and some kind of direction of effect for nearly all (>96%). If we select either of the larger tables, we get evidence for about half of all clincial annotations.

#### Direction of effect representation

If we focus on only one of these tables, we need "only" go through an exercise of selecting which columns we want to extract, and which (if any) we want to map. If we want to use all three tables, since the column schema differs among them, we could do the same thing and have lots of optional attributes to cover all three types of annotation, or just use the free-text sentence that is provided for each annotation.

If we want to use all three tables but present them in a unified and structured way, we need to come up with a generic representation of what "direction of effect" means. Based on the [sentence breakdown](https://nbviewer.org/github/apriltuesday/opentargets-pharmgkb/blob/doe/data-exploration/direction-of-effect.ipynb?flush_cache=true#Sentence-breakdown) I came up with one suggestion, but this obviously would require much more discussion:

1. Direction of effect (`decreased`, `increased`, or none)
2. PD/PK term | Functional term | Side effect/efficacy/other (e.g. `response to`, `activity of`, `severity of`)
3. Drug | Gene/gene product | Phenotype (e.g. `ivacaftor`, `CFTR`, `side effect: bone density`)

This tells us the direction (1) and what the effect is (2&3). Of course we need some other fields to connect things, but this could be the core "direction of effect" concept.

Of the above, (1) is always either "decreased" or "increased", (2) takes values in a relatively small but not fixed vocabulary (could perhaps be mapped to an EFO term), and (3) is open and would probably ideally be mapped. The values that appear can be found [here](https://nbviewer.org/github/apriltuesday/opentargets-pharmgkb/blob/doe/data-exploration/direction-of-effect.ipynb?flush_cache=true#Vocabulary).

#### Allele / genotype representation

This is really about how to connect the direction of effect to our current evidence strings. Associating variant annotations to clinical annotations via PMID or evidence ID is the most straightfoward method. Logically, however, direction of effect annotations should be allele / genotype specific, so I looked briefly into how these are represented in the new tables - basically, can we get a direction of effect per genotype or haplotype ID.

I think this is mostly doable, as the representation is consistent with what we've seen in the clinical annotations tables. The exception is that sometimes they've provided a metabolyzer type instead of an allele/genotype for comparison in the variant annotation. See [here](https://nbviewer.org/github/apriltuesday/opentargets-pharmgkb/blob/doe/data-exploration/direction-of-effect.ipynb?flush_cache=true#Alleles-and-genotypes) for an example. Honestly I have no idea how we could handle these right now.

#### Some questions to consider
* Which tables do we want to include?
    * Basically are we interested in all the types of effects described
* Which columns do we want to use? (for v1 at least)
    * Direction of effect can be quite a complex concept, what parts do we want to use as-is (strings) vs. map to ids vs. omit entirely?
* How do we want to make the association with clinical annotations?
    * via allele/genotype, via PMID, something else?
    * aggregated (as PMID is currently) or exploded (as allele/genotype)?

**Comment by apriltuesday** [comment]
As discussed, [this spreadsheet](https://docs.google.com/spreadsheets/d/1NIgVnQ0XHEwW6U_lTAoXnW_TYoL-2d_WDs_vwcoSGa8/edit?usp=sharing) contains examples of clinical annotations with their allele/genotype annotations and variant annotations. There are 4 clinical annotations in total, including one with "metabolyzer type" comparisons which we didn't have time to cover in the meeting. The data is split into tabs by variant annotation type and includes all columns, though I've hidden some to make things a bit easier to read.

Let me know any questions or thoughts you have!

**Comment by apriltuesday** [comment]
@tcezard @DSuveges @ireneisdoomed @tskir - I've been looking into how to associate variant annotations with clinical annotations at the level of genotype/allele. We can go over this together in our next meeting, but meanwhile feel free to leave your thoughts and questions.

I've done a rough proof-of-concept of what the associations might look like. The algorithm is not particularly clever, it basically amounts to decomposing genotype/allele strings until we get to alleles, and doing exact string matching to line things up. It also makes a core assumption, which is that **annotations on alleles can be associated with any genotype containing that allele**. So if we have variant annotations on alleles `C` and `T`, both will be associated with genotype `CT`. Conversely, if we have a variant annotation on genotype `*1/*18`, it will be associated with alleles `*1` and `*18`.

You can see the algorithm itself in the notebook [here](https://nbviewer.org/github/apriltuesday/opentargets-pharmgkb/blob/variant-clinical-association/data-exploration/variant-clinical-annotation-association.ipynb?flush_cache=true) and the results on a handful of examples in the spreadsheet [here](https://docs.google.com/spreadsheets/d/1ntxUThzoF80W2rn17oW-kYvtw1jsBZxWcZN427qh3Yk/edit?usp=sharing) (with columns removed for readability).

I also ran this on the entire dataset and found it could successfully associate **93.7%** of variant annotations, and it found at least one variant annotation for **65.1%** of clinical annotation _genotypes_. There are definitely some tricky cases I know it fails on, and probably some I don't know about! Also, the approximately 2/3 clinical annotation genotype coverage is not unexpected given that we often don't have variant annotations for the ref/ref genotype.

I've got at least a couple questions for us to discuss:
1. Does the assumption highlighted above (about genotypes/allele associations) seem reasonable?
2. What we should do about ref/ref genotypes?

**Comment by tcezard** [comment]
I think we should start by reiterating what we are aiming to get out of the association between genotype and variant annotation: 
 - A set of decomposed statement for each genotypes that are hopefully concordant and can be summarised programatically (assuming all the statments have the same direction of effects)
 - A set of statements associated with each genotype that can be fed to chat-gpt to help summarise the direction of effects
 - A set of publications associated with each genotype to give confidence in each statement as they could provide independent sources.

This being said, the algorithm is good enough for a first pass and without starting to explode the * alleles in their individual rs components we might not get much better.

For the ref/ref genotypes, they should be associated with the evidence that mentions this allele and not the other. OT could on its side highlight this genotype as the ref/ref so not associated with any evidence

I though we said previously that all Clinical annotation had at least one piece of evidence associated from either of the 3 source. Is the 65% due to the fact that we are counting the exploded clinical annotation per genotype id ?

**Comment by apriltuesday** [comment]
> I though we said previously that all Clinical annotation had at least one piece of evidence associated from either of the 3 source. Is the 65% due to the fact that we are counting the exploded clinical annotation per genotype id ?

Yes that's exactly it. I just checked coverage of clinical annotations not exploded by genotype, and it's **99.3%**. (Of course if we just wanted to associate variant annotations with clinical annotations, not exploding by allele/genotype, our coverage would be 100% as we can do this directly by IDs...)

**Comment by apriltuesday** [comment]
@DSuveges @ireneisdoomed Hi all, based on past discussions I've looked a bit into how we might extract or summarise direction of effect information from associated variant annotations. I've written this up in a [document](https://docs.google.com/document/d/1YYNMLArt0FNFUEFLieMDm1p5NHSfJheddnPUocy8pGM/edit?usp=sharing) along with some examples of what the evidence strings could look like, just as a first step. Please take a look and leave your comments.

**Comment by apriltuesday** [comment]
PR #50 adds a feature flag for including direction of effect in evidence strings for the purpose of generating a complete test set, will iterate on this before merging and including in future submissions.

**Comment by ireneisdoomed** [comment]
I have posted in the PR and prob it is more convenient here:

> Hi @apriltuesday,
After discussion, we would like to have the information that combines the directionality split by columns so it is easier to analyse the level of agreement between all variation for a given clinical annotation.
This would be adding directionality, effect, and entity into the evidenceFromSource. struct. Is it possible to put together?
Thanks!

**Comment by apriltuesday** [comment]
Hi @ireneisdoomed , yes that's fine, I will update the code. Did you have any thoughts on the `evidenceFromSource` name?

**Comment by ireneisdoomed** [comment]
> Did you have any thoughts on the evidenceFromSource name?

Not really. I thought about `variantAnnotation` but we haven't decided. It's not important for the time being

**Comment by ireneisdoomed** [comment]
Do you think that is covered by `pgxCategory` from the clinical annotations? Maybe something to check is that the type of variant annotation is the same in all the evidence records (and it matches the clinical annotation)

**Comment by apriltuesday** [comment]
It's not the same as `pgxCategory`, and some clinical annotations will have multiple types of variant annotations. An example that has all three types is [this one](https://www.pharmgkb.org/clinicalAnnotation/981755803).  (I'm not sure where on the website it states the type of variant annotation, but it's clear in the data itself - I got this example by looking at [this spreadsheet of examples](https://docs.google.com/spreadsheets/d/1NIgVnQ0XHEwW6U_lTAoXnW_TYoL-2d_WDs_vwcoSGa8/edit?usp=sharing).)

**Comment by ireneisdoomed** [comment]
@apriltuesday Thank you for the example. Could you please include it? My initial thought is that the category should be homogenous, and perhaps we should separate them, but I want to see if there are more extreme examples where the variant annotation reports cases of toxicity but the category of the clinical annotation is Efficacy/Dosage. Thanks!

**Comment by apriltuesday** [comment]
Addressed in #55
